[HIP][Clang][CodeGen] Handle hip bin symbols properly.

jofrn opened the pull request:
Remove '_' in fatbin and gpubin symbol suffixes when missing TU hash ID. Internalize gpubin symbol so that it is not unresolved at link-time when symbol is not relocatable.

yxsamliu commented:
The current behavior of clang is expected.

when gpu binary is not specified, it is expected to be used for -fgpu-rdc and the __hip_gpubin_handle_ symbol needs to be external and unique since they may need to be merged for partial linking. Make them internal will break partial linking.

yxsamliu commented on clang/lib/CodeGen/CGCUDANV.cpp:
```diff
@@ -894,8 +896,7 @@ llvm::Function *CGNVCUDARuntime::makeModuleCtorFunction() {
   if (IsHIP) {
-    auto Linkage = CudaGpuBinary ? llvm::GlobalValue::InternalLinkage
-                                 : llvm::GlobalValue::ExternalLinkage;
+    auto Linkage = llvm::GlobalValue::InternalLinkage;
```
CudaGpuBinary is not always true.

For normal `-fgpu-rdc` usage, CudaGpuBinary is false here.

I think the proper fix is 

```
auto Linkage = RelocatableDeviceCode? llvm::GlobalValue::ExternalLinkage
                                 : llvm::GlobalValue::InternalLinkage;

```

yxsamliu commented on clang/lib/CodeGen/CGCUDANV.cpp:
```diff
@@ -910,9 +911,6 @@ llvm::Function *CGNVCUDARuntime::makeModuleCtorFunction() {
     GpuBinaryHandle->setAlignment(CGM.getPointerAlign().getAsAlign());
-    // Prevent the weak symbol in different shared libraries being merged.
-    if (Linkage != llvm::GlobalValue::InternalLinkage)
-      GpuBinaryHandle->setVisibility(llvm::GlobalValue::HiddenVisibility);
```
should be kept

yxsamliu commented on clang/lib/CodeGen/CGCUDANV.cpp:
```diff
@@ -840,8 +840,10 @@ llvm::Function *CGNVCUDARuntime::makeModuleCtorFunction() {
-          "__hip_fatbin_" + CGM.getContext().getCUIDHash(), nullptr,
-          llvm::GlobalVariable::NotThreadLocal);
+          "__hip_fatbin" + (CGM.getLangOpts().CUID.empty()
+                                ? ""
```
this change is unnecessary and may break some existing app

jofrn replied:
The apps that relied on the ID without an '_' broke to begin with, so we should go ahead with it.(?)

jofrn commented:
Thank you for your comments. Updated the code to use 'RelocatableDeviceCode' in the case of external linkage.

yxsamliu commented:
you need to update lit test clang/test/CodeGenCUDA/device-stub.cu as it fails now

[clang/test/CodeGenCUDA/device-stub.cu](https://buildkite.com/llvm-project/github-pull-requests/builds/98392#0191c73e-12f8-4cf3-9618-d3fd752f9149)

yxsamliu commented on clang/lib/CodeGen/CGCUDANV.cpp:
```diff
@@ -905,10 +907,10 @@ llvm::Function *CGNVCUDARuntime::makeModuleCtorFunction() {
     GpuBinaryHandle = new llvm::GlobalVariable(
         TheModule, PtrTy, /*isConstant=*/false, Linkage,
         /*Initializer=*/
-        CudaGpuBinary ? llvm::ConstantPointerNull::get(PtrTy) : nullptr,
```
this line should not be changed

yxsamliu replied:
or change it to:

`RelocatableDeviceCode ? nullptr : llvm::ConstantPointerNull::get(PtrTy)`

jofrn replied:
Yes, that one is good! Thanks.

yxsamliu commented on clang/test/CodeGenCUDA/device-stub.cu:
```diff
@@ -175,7 +175,6 @@ __device__ void device_use() {
 // HIP-SAME: section ".hipFatBinSegment"
 // * variable to save GPU binary handle after initialization
 // CUDANORDC: @__[[PREFIX]]_gpubin_handle = internal global ptr null
-// HIPNEF: @__[[PREFIX]]_gpubin_handle_{{[0-9a-f]+}} = external hidden global ptr, align 8
```
should not remove the check line

jofrn committed:
[HIP][Clang][CodeGen] Handle hip bin symbols properly.

Remove '_' in fatbin symbol suffix when missing TU hash ID. Internalize gpubin symbol so that it is not unresolved at link-time.

yxsamliu reviewed (approved):
LGTM. Thanks
